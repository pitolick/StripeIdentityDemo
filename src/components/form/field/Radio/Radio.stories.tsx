import type { Meta, StoryObj } from '@storybook/react'
import { Radio } from './Radio'

const meta: Meta<typeof Radio> = {
  title: 'Form/Field/Radio/Base',
  component: Radio,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof Radio>

/**
 * 基本形
 */
export const Base: Story = {
  args: {},
}

/**
 * ラベル
 */
export const Label: Story = {
  args: {
    labelContents: 'ラベル',
  },
}

/**
 * ラベルDOM
 */
export const LabelDom: Story = {
  args: {
    labelContents: <span>ラベル</span>,
  },
}

/**
 * className
 */
export const ClassName: Story = {
  args: {
    radioClassName: 'border-2',
    labelClassName: 'inline-flex items-center space-x-2',
    labelContents: <span>ラベル</span>,
  },
}

/**
 * デフォルトチェック
 */
export const DefaultChecked: Story = {
  args: {
    defaultChecked: true,
    labelContents: <span>ラベル</span>,
  },
}

/**
 * ラッパー：true
 */
export const Wrapper: Story = {
  args: {
    wrapper: true,
    labelContents: <span>ラベル</span>,
  },
}

/**
 * ラッパー：false
 */
export const NoWrapper: Story = {
  args: {
    wrapper: false,
    labelContents: <span>ラベル</span>,
  },
}
