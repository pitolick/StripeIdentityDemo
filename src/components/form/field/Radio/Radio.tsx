/**
 * ラジオボタン非制御コンポーネント
 * @module Radio
 */
import React, { forwardRef } from 'react'

/**
 * ラジオボタンコンポーネントのプロパティ
 * @typedef {Object} RadioProps
 * @property {React.InputHTMLAttributes<HTMLInputElement>} - ラジオボタンタグのHTML属性
 */
export interface RadioProps extends React.InputHTMLAttributes<HTMLInputElement> {
  /**
   * class名：ラジオボタン
   */
  radioClassName?: React.HTMLAttributes<HTMLInputElement>['className']
  /**
   * class名：ラベル
   */
  labelClassName?: React.HTMLAttributes<HTMLLabelElement>['className']
  /**
   * ラベルコンテンツ
   */
  labelContents?: React.ReactNode
  /**
   * ラベル（確認画面表示用）
   */
  label?: string
  /**
   * ラッパー
   */
  wrapper?: boolean
}

/**
 * ラジオボタン非制御コンポーネント
 * @param {RadioProps} props - ラジオボタンコンポーネントのプロパティ
 * @returns {JSX.Element} - ラジオボタンコンポーネント
 */
export const Radio = forwardRef<HTMLInputElement, RadioProps>((props, ref) => {
  const { radioClassName = undefined, labelClassName = undefined, labelContents = undefined, wrapper = true, ...radioProps } = props
  const radioClassNameProps = radioClassName ? radioClassName : ''
  const labelClassNameProps = labelClassName ? labelClassName : ''

  // ラッパーコンポーネント
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    // wrapperがtrueの場合はlabelをfalseの場合はフラグメントを返す
    return wrapper ? <label className={`${labelClassNameProps} cursor-pointer`}>{children}</label> : <>{children}</>
  }

  return (
    <>
      <Wrapper>
        <input ref={ref} {...radioProps} type="radio" className={`focus:ring-transparent ${radioClassNameProps}`} />
        {labelContents && <span className="text-sm md:text-base">{labelContents}</span>}
      </Wrapper>
    </>
  )
})

Radio.displayName = 'Radio'
