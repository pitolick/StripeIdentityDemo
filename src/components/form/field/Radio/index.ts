export * from './Radio'
export * from './RadioControl'
