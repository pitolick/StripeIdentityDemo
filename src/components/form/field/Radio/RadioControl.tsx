/**
 * ラジオボタン制御コンポーネント（react-hook-form）
 * @module RadioControl
 * @todo カードスタイルのときにTabでフォーカスした場合にアウトラインがでないのを修正
 */
import React, { forwardRef } from 'react'
import { Radio, RadioProps } from './Radio'
import { InputBox, InputBoxProps } from '@/components/form/parts'

/**
 * ラジオボタンコンポーネントのプロパティ
 * @typedef {Object} RadioControlProps
 * @property {React.InputHTMLAttributes<HTMLInputElement>} - ラジオボタンタグのHTML属性
 */

export interface RadioControlProps extends React.InputHTMLAttributes<HTMLInputElement>, Omit<InputBoxProps, 'children'> {
  /**
   * リスト
   */
  lists: RadioProps[]
  /**
   * ラベルのタイプ
   */
  labelType?: 'default'
  /**
   * 並び方
   */
  direction?: 'row' | 'col'
  /**
   * ラジオボタンの位置
   */
  radioPosition?: 'center' | 'top' | 'hidden'
  /**
   * ラジオボタンの色
   */
  radioColor?: 'black' | 'red' | 'blue' | 'gray'
}

/**
 * ラジオボタン制御コンポーネント（react-hook-form）
 * @param {RadioControlProps} props - ラジオボタンコンポーネントのプロパティ
 * @returns {JSX.Element} ラジオボタンコンポーネント
 */
export const RadioControl = forwardRef<HTMLInputElement, RadioControlProps>((props, ref) => {
  const { lists, nameId, label, note, required = undefined, error, labelType = 'default', direction = 'row', radioPosition = 'center', radioColor = 'blue', contentWidth, ...radioControlProps } = props

  const radioList = lists.map((list, index) => {
    const { labelContents, radioClassName, labelClassName, ...radioProps } = list
    // class名
    let radioClassNameProps = `${radioClassName ? radioClassName : ''}`
    let labelClassNameProps = `${labelClassName ? labelClassName : ''} inline-block relative pl-5`

    // ラジオボタンの位置
    switch (radioPosition) {
      case 'center':
        radioClassNameProps = radioClassNameProps + ' absolute margin-auto top-1/2 transform -translate-y-1/2'
        break
      case 'top':
        radioClassNameProps = radioClassNameProps + ' absolute margin-auto'
        // ラベルのタイプによって分岐
        switch (labelType) {
          case 'default':
            radioClassNameProps = radioClassNameProps + ' top-1'
            break
        }
        break
      case 'hidden':
        radioClassNameProps = 'hidden'
        break
    }

    // エラー時のclass名
    let errorLabelClass = ''
    let errorRadioClass = ''

    // ラベルのタイプ
    switch (labelType) {
      case 'default':
        radioClassNameProps = radioClassNameProps + ' left-0'
        errorLabelClass = error ? ' text-red' : ''
        errorRadioClass = error ? ' bg-red-pale text-red border-red' : ''
        break
    }

    // ラジオボタンの色
    switch (radioColor) {
      case 'black':
        radioClassNameProps = radioClassNameProps + ' checked:text-black focus:text-black'
        break
      case 'red':
        radioClassNameProps = radioClassNameProps + ' checked:text-red focus:text-red'
        break
      case 'blue':
        radioClassNameProps = radioClassNameProps + ' checked:text-blue focus:text-blue'
        break
      case 'gray':
        radioClassNameProps = radioClassNameProps + ' checked:text-gray-700 focus:text-gray-700'
        break
    }

    // カードコンポーネント
    const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
      return <>{children}</>
    }

    return (
      <div key={`${nameId}_${index}`} className={`${direction === 'col' && 'w-full'} ${direction === 'row' && contentWidth === 'full' && 'flex-auto'}`}>
        <Wrapper>
          <Radio
            ref={ref}
            labelContents={labelContents}
            radioClassName={`${radioClassNameProps} ${errorRadioClass}`}
            labelClassName={`${labelClassNameProps} ${errorLabelClass}`}
            wrapper={true}
            {...radioProps}
            {...radioControlProps}
          />
        </Wrapper>
      </div>
    )
  })

  return (
    <InputBox nameId={nameId} label={label} note={note} required={required} error={error} contentWidth={contentWidth}>
      {/* ラジオボタン */}
      <div className={`${'flex-' + direction} ${contentWidth === 'full' && 'w-full'} flex flex-wrap gap-3`}>{radioList}</div>
    </InputBox>
  )
})

RadioControl.displayName = 'RadioControl'
