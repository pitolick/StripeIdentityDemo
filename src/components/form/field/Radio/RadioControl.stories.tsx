import type { Meta, StoryObj } from '@storybook/react'
import { RadioControl } from './RadioControl'

const meta: Meta<typeof RadioControl> = {
  title: 'Form/Field/Radio/Control',
  component: RadioControl,
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof RadioControl>

/**
 * 基本形
 */
export const Base: Story = {
  args: {
    nameId: 'base',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', name: 'base' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'base' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'base' },
    ],
  },
}

/**
 * 基本形：デフォルトチェック（複数ある場合は最後の要素がチェックされる）
 */
export const BaseDefaultChecked: Story = {
  args: {
    nameId: 'base-default-checked',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', defaultChecked: true, name: 'base-default-checked' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'base-default-checked' },
      { labelContents: 'ラジオボタン3', value: '3', defaultChecked: true, name: 'base-default-checked' },
    ],
  },
}

/**
 * 基本形：DOM
 */
export const BaseDom: Story = {
  args: {
    nameId: 'base-dom',
    lists: [
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン1
          </span>
        ),
        value: '1',
        name: 'base-dom',
      },
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン2
          </span>
        ),
        value: '2',
        name: 'base-dom',
      },
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン3
          </span>
        ),
        value: '3',
        name: 'base-dom',
      },
    ],
  },
}

/**
 * className
 */
export const ClassName: Story = {
  args: {
    nameId: 'class-name',
    lists: [
      {
        labelContents: 'ラジオボタン1',
        value: '1',
        name: 'class-name',
        radioClassName: 'border-4',
        labelClassName: 'text-5xl',
      },
      {
        labelContents: <span className="text-4xl">ラジオボタン2</span>,
        value: '2',
        name: 'class-name',
      },
      { labelContents: 'ラジオボタン3', value: '3', name: 'class-name' },
    ],
  },
}

/**
 * ラベル
 */
export const Label: Story = {
  args: {
    nameId: 'label',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', name: 'label' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'label' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'label' },
    ],
    label: 'ラベル',
  },
}

/**
 * バリデーション：必須
 */
export const ValidationRequired: Story = {
  args: {
    nameId: 'validation-required',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', name: 'validation-required' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'validation-required' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'validation-required' },
    ],
    label: 'ラベル',
    required: true,
  },
}

/**
 * バリデーション：任意
 */
export const ValidationOptional: Story = {
  args: {
    nameId: 'validation-optional',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', name: 'validation-optional' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'validation-optional' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'validation-optional' },
    ],
    label: 'ラベル',
    required: false,
  },
}

/**
 * バリデーション：エラー
 */
export const ValidationError: Story = {
  args: {
    nameId: 'error',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', defaultChecked: true, name: 'error' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'error' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'error' },
    ],
    label: 'ラベル',
    required: true,
    error: { type: 'required', message: '必須項目です' },
  },
}

/**
 * 注釈
 */
export const Note: Story = {
  args: {
    nameId: 'note',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', name: 'note' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'note' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'note' },
    ],
    label: 'ラベル',
    note: { list: ['注釈'] },
  },
}

/**
 * 並び方：横
 */
export const DirectionRow: Story = {
  args: {
    nameId: 'direction-row',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', name: 'direction-row' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'direction-row' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'direction-row' },
      { labelContents: 'ラジオボタン4', value: '4', name: 'direction-row' },
      { labelContents: 'ラジオボタン5', value: '5', name: 'direction-row' },
      { labelContents: 'ラジオボタン6', value: '6', name: 'direction-row' },
      { labelContents: 'ラジオボタン7', value: '7', name: 'direction-row' },
      { labelContents: 'ラジオボタン8', value: '8', name: 'direction-row' },
      { labelContents: 'ラジオボタン9', value: '9', name: 'direction-row' },
    ],
    label: 'ラベル',
    direction: 'row',
  },
}

/**
 * 並び方：縦
 */
export const DirectionColmun: Story = {
  args: {
    nameId: 'direction-colmun',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', name: 'direction-colmun' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'direction-colmun' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'direction-colmun' },
    ],
    label: 'ラベル',
    direction: 'col',
  },
}

/**
 * ラジオボタンの位置：中央
 */
export const RadioPosition: Story = {
  args: {
    nameId: 'radio-position',
    lists: [
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン1
          </span>
        ),
        value: '1',
        name: 'radio-position',
      },
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン
            <br />2
          </span>
        ),
        value: '2',
        name: 'radio-position',
      },
      { labelContents: <span>ラジオボタン3</span>, value: '3', name: 'radio-position' },
    ],
    label: 'ラベル',
    radioPosition: 'center',
  },
}

/**
 * ラジオボタンの位置：上
 */
export const RadioPositionTop: Story = {
  args: {
    nameId: 'radio-position-top',
    lists: [
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン1
          </span>
        ),
        value: '1',
        name: 'radio-position-top',
      },
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン
            <br />2
          </span>
        ),
        value: '2',
        name: 'radio-position-top',
      },
      { labelContents: <span>ラジオボタン3</span>, value: '3', name: 'radio-position-top' },
    ],
    label: 'ラベル',
    radioPosition: 'top',
  },
}

/**
 * ラジオボタンの位置：非表示
 */
export const RadioPositionHidden: Story = {
  args: {
    nameId: 'radio-position-hidden',
    lists: [
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン1
          </span>
        ),
        value: '1',
        name: 'radio-position-hidden',
      },
      {
        labelContents: (
          <span>
            ラジオ
            <br />
            ボタン
            <br />2
          </span>
        ),
        value: '2',
        name: 'radio-position-hidden',
      },
      { labelContents: <span>ラジオボタン3</span>, value: '3', name: 'radio-position-hidden' },
    ],
    label: 'ラベル',
    radioPosition: 'hidden',
  },
}

/**
 * ラジオボタンの色：黒
 */
export const RadioColorBlack: Story = {
  args: {
    nameId: 'radio-color-black',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', defaultChecked: true, name: 'radio-color-black' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'radio-color-black' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'radio-color-black' },
    ],
    label: 'ラベル',
    radioColor: 'black',
  },
}

/**
 * ラジオボタンの色：赤
 */
export const RadioColorRed: Story = {
  args: {
    nameId: 'radio-color-red',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', defaultChecked: true, name: 'radio-color-red' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'radio-color-red' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'radio-color-red' },
    ],
    label: 'ラベル',
    radioColor: 'red',
  },
}

/**
 * ラジオボタンの色：青
 */
export const RadioColorBlue: Story = {
  args: {
    nameId: 'radio-color-blue',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', defaultChecked: true, name: 'radio-color-blue' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'radio-color-blue' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'radio-color-blue' },
    ],
    label: 'ラベル',
    radioColor: 'blue',
  },
}

/**
 * ラジオボタンの色：グレー
 */
export const RadioColorGray: Story = {
  args: {
    nameId: 'radio-color-gray',
    lists: [
      { labelContents: 'ラジオボタン1', value: '1', defaultChecked: true, name: 'radio-color-gray' },
      { labelContents: 'ラジオボタン2', value: '2', name: 'radio-color-gray' },
      { labelContents: 'ラジオボタン3', value: '3', name: 'radio-color-gray' },
    ],
    label: 'ラベル',
    radioColor: 'gray',
  },
}
