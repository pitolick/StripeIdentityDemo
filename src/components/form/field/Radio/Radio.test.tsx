import React from 'react'
import { render } from '@testing-library/react'
import { Radio } from './Radio'

describe('Radio', () => {
  it('正しくレンダリングされること', () => {
    const { container } = render(<Radio id="my-radio" label="ラジオ" labelContents="ラベルコンテンツ" radioClassName="custom-radio" labelClassName="custom-label" wrapper={true} />)
    expect(container).toMatchSnapshot()
  })

  it('デフォルト値が正しく設定されること', () => {
    const { container } = render(<Radio id="my-radio" />)
    expect(container).toMatchSnapshot()
  })
})
