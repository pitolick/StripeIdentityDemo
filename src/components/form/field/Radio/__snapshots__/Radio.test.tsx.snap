// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio デフォルト値が正しく設定されること 1`] = `
<div>
  <label
    class=" cursor-pointer"
  >
    <input
      class="focus:ring-transparent "
      id="my-radio"
      type="radio"
    />
  </label>
</div>
`;

exports[`Radio 正しくレンダリングされること 1`] = `
<div>
  <label
    class="custom-label cursor-pointer"
  >
    <input
      class="focus:ring-transparent custom-radio"
      id="my-radio"
      label="ラジオ"
      type="radio"
    />
    <span
      class="text-sm md:text-base"
    >
      ラベルコンテンツ
    </span>
  </label>
</div>
`;
